lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,199 @@
+#     * `Float` - floating point number (same signature as for Integer for `Integer` ranges)
+#     * `Boolean` - true of false (strict)
+#     * `Array` - an array (of Data by default), or parameterized as `Array[<element_type>]`, where
+#       `<element_type>` is the expected type of elements
+#     * `Hash`,  - a hash (of default `Literal` keys and `Data` values), or parameterized as
+#       `Hash[<value_type>]`, `Hash[<key_type>, <value_type>]`, where `<key_type>`, and
+#       `<value_type>` are the types of the keys and values respectively
+#       (key is `Literal` by default).
+#     * `Data` - abstract type representing any `Literal`, `Array[Data]`, or `Hash[Literal, Data]`
+#     * `Pattern[<p1>, <p2>, ..., <pn>]` - an enumeration of valid patterns (one or more) where
+#        a pattern is a regular expression string or regular expression,
+#        e.g. `Pattern['.com$', '.net$']`, `Pattern[/[a-z]+[0-9]+/]`
+#     * `Enum[<s1>, <s2>, ..., <sn>]`, - an enumeration of exact string values (one or more)
+#        e.g. `Enum[blue, red, green]`.
+#     * `Variant[<t1>, <t2>,...<tn>]` - matches one of the listed types (at least one must be given)
+#       e.g. `Variant[Integer[8000,8999], Integer[20000, 99999]]` to accept a value in either range
+#     * `Regexp`- a regular expression (i.e. the result is a regular expression, not a string
+#        matching a regular expression).
+#   * A string containing a type description - one of the types as shown above but in string form.
+#
+# If the function is called without specifying a default value, and nothing is bound to the given name
+# an error is raised unless the option `accept_undef` is true. If a block is given it must produce an acceptable
+# value (or call `error`). If the block does not produce an acceptable value an error is
+# raised.
+#
+# Examples:
+#
+# When called with one argument; **the name**, it
+# returns the bound value with the given name after having  asserted it has the default datatype `Data`:
+#
+#     lookup('the_name')
+#
+# When called with two arguments; **the name**, and **the expected type**, it
+# returns the bound value with the given name after having asserted it has the given data
+# type ('String' in the example):
+#
+#     lookup('the_name', 'String')
+#     lookup('the_name', String)
+#
+# When called with three arguments, **the name**, the **expected type**, and a **default**, it
+# returns the bound value with the given name, or the default after having asserted the value
+# has the given data type (`String` in the example above):
+#
+#     lookup('the_name', 'String', 'Fred')
+#     lookup('the_name', String, 'Fred')
+#
+# Using a lambda to provide a default value by calling a function:
+#
+#     lookup('the_size', Integer[1,100]) |$name| {
+#       obtain_size_default()
+#     }
+#
+# When using a block, the value it produces is also asserted against the given type, and it may not be
+# `undef` unless the option `'accept_undef'` is `true`.
```

Comment:
There was language above that also said things about undef (slightly confusing there, more clear how it works here).